empty-bin.lua

```diff
@@ -3,24 +3,51 @@
 -- https://gist.github.com/stonetoad/11129025
 -- http://dwarffortresswiki.org/index.php/DF2014_Talk:Bin
 
---[====[
-
-empty-bin
-=========
+local function moveItem(item, to_pos)
+    print('  ' .. dfhack.items.getDescription(item, 0))
+    dfhack.items.moveToGround(item, to_pos)
+end
 
-Empties the contents of the selected bin onto the floor.
+local function emptyContainer(container)
+    local items = dfhack.items.getContainedItems(container)
 
-]====]
+    if #items > 0 then
+        print('Emptying ' .. dfhack.items.getDescription(container, 0))
+        for _, item in ipairs(items) do
+            moveItem(item, xyz2pos(dfhack.items.getPosition(container)))
+        end
+    end
+end
 
-local bin = dfhack.gui.getSelectedItem(true) or qerror("No item selected")
-local items = dfhack.items.getContainedItems(bin)
+local viewsheets = df.global.game.main_interface.view_sheets
 
-if #items > 0 then
-    print('Emptying ' .. dfhack.items.getDescription(bin, 0))
-    for _, item in pairs(items) do
-        print('  ' .. dfhack.items.getDescription(item, 0))
-        dfhack.items.moveToGround(item, xyz2pos(dfhack.items.getPosition(bin)))
+local stockpile = dfhack.gui.getSelectedStockpile(true)
+local selectedItem = dfhack.gui.getSelectedItem(true)
+local selectedBuilding = dfhack.gui.getSelectedBuilding(true)
+if stockpile then
+    local contents = dfhack.buildings.getStockpileContents(stockpile)
+    for _, container in ipairs(contents) do
+        emptyContainer(container)
+    end
+elseif selectedItem then
+    moveItem(selectedItem, xyz2pos(dfhack.items.getPosition(selectedItem)))
```

Comment:
what is the intention of this line? why move the item itself?

Reply:
If you select a specific item that's already inside a container (perhaps you want to dump *just* that item), this code will properly move just that item to the ground